add image prep functions to imghelper

jpculp opened the pull request:
**Description of changes:**

Adds or moves functionality around image name generation, compression, and decompression to `imghelper`.

**Testing done:**

- Built `aws-k8s-1.28`
- Repacked `aws-k8s-1.28`
- Build `vmware-k8s-1.28`
- Repacked `vmware-k8s-1.28`



**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

webern commented on twoliter/embedded/imghelper:
```diff
@@ -1,6 +1,150 @@
+
+    if [ "${partition_plan}" == "unified" ]; then
+      if grep -Fq '{{DATA_DISK}}' "${ovf_template}"; then
+        echo "Incorrect data disk in OVF template, which is not supported for 'unified' layout." >&2
```
?

```suggestion
        echo "Data disk included in OVF template, which is not supported for 'unified' layout." >&2
```

bcressey commented on twoliter/embedded/imghelper:
```diff
@@ -1,6 +1,150 @@
+
+  if [[ "${what}" == "os_image" ]] || [[ "${what}" == "data_image" ]]; then
+    if [[ "${ext}" == "img.lz4" ]]; then
+      lz4 -vc "${!input_image}" >"${output_dir}/${!image_name}.img.lz4"
```
It looks like this would be more readable with a couple changes:
* always append `${ext}` to `${!image_name}` when writing the output (this would work for boot/root/hash also where the empty string is passed)
* always use `lz4 -9vc` for "img.lz4" - not sure if this really saves much time for the os image; it's important for the boot/root/hash images to stay small.

Then this section can be just a case statement:

```
case "${ext}" in
  ""|.img.lz4)
    lz4 -9vc "${!input_image}" \
      > "${output_dir}/${!image_name}${ext:+.${ext}}"
    ;;
  qcow2)
    ...
  vmdk)
    ...
  *)
    echo "unexpected extension: ${ext}" >&2
    exit 1
    ;;
esac
```

bcressey commented on twoliter/embedded/imghelper:
```diff
@@ -1,6 +1,150 @@
+  fi
+}
+
+decompress_ova() {
+  local fmt input_ova filename_prefix output_image
+  fmt="${1:?}"
```
It feels like for consistency with `compress_image`, this should be a supported extension in `decompress_image` rather than a separate function.

jpculp replied:
I originally split this up because the we vend OVAs with the "friendly" name as opposed to the "versioned" name for our raw images. However, since we're going to generate missing symlinks in the end anyway, I can just create a symlink for the ova at the start.

bcressey commented on twoliter/embedded/imghelper:
```diff
@@ -1,6 +1,150 @@
+
+  tar xvf "${input_ova}" "${filename_prefix}.vmdk"
+  qemu-img convert -f vmdk -O raw "${filename_prefix}.vmdk" "${output_image}"
+  rm "${filename_prefix}.vmdk"
```
With a split VMDK, we might end up with two images to extract.

If you take an output directory, rather than a filename, this becomes easier to do - just extract files matching `*.vmdk` from the OVA. It seems like the function caller already knows what images to expect (because of `set_image_names`).

bcressey commented on twoliter/embedded/imghelper:
```diff
@@ -1,6 +1,150 @@
-
 # shellcheck disable=SC2034 # variables are used through parameter substitution
+
+DATA_IMAGE_SUFFIX="-data"
+BOOT_IMAGE_SUFFIX="-boot.ext4.lz4"
+VERITY_IMAGE_SUFFIX="-root.verity.lz4"
+ROOT_IMAGE_SUFFIX="-root.ext4.lz4"
```
It's weird that the data image variable doesn't have an extension but the other three do. It suggests they're not really being treated the same way and probably shouldn't share a `_SUFFIX` suffix.

jpculp replied:
Another weird behavior inherited from `rpm2img`. It was probably written this way since the extension for data depends on the build args whereas the others are constant. We can drop them and add the extensions to the `symlink_image` calls, it just means a bigger delta from the original script.

bcressey commented on twoliter/embedded/imghelper:
```diff
@@ -1,6 +1,150 @@
+  fi
+}
+
+get_data_image() {
+  local fmt input_dir filename_prefix
+  fmt="${1:?}"
+  input_dir="${2:?}"
+  filename_prefix="${3:?}"
+
+  if [[ "${fmt}" == "raw" ]]; then
+    echo "${input_dir}/${filename_prefix}${DATA_IMAGE_SUFFIX}.img.lz4"
+  elif [[ "${fmt}" == "qcow2" ]]; then
+    echo "${input_dir}/${filename_prefix}${DATA_IMAGE_SUFFIX}.qcow2"
+  elif [[ "${fmt}" == "vmdk" ]]; then
+    echo "${input_dir}/${filename_prefix}${DATA_IMAGE_SUFFIX}.vmdk"
+  fi
+}
```
Without seeing the caller for this function, it's hard to review but it does seem a bit strange. The caller has almost all of the information (because `${filename_prefix}${DATA_IMAGE_SUFFIX}` is the same as the `data_image_name` we set for them.

jpculp replied:
I could save this for the `img2img` PR, but weirdness comes from our choice of formats. "raw" == "img.lz4", so we can't just do `${INPUT_DIR}/${FILENAME_PREFIX}${DATA_IMAGE_SUFFIX}.${OUTPUT_FMT}`. I could do `${INPUT_DIR}/${FILENAME_PREFIX}${DATA_IMAGE_SUFFIX}.${OUTPUT_FMT/raw/img.lz4}`, but ideally filename prefix and data image suffix will be in imghelper and not the caller. This seemed the most readable when factoring in the above.

bcressey commented on twoliter/embedded/imghelper:
```diff
@@ -1,6 +1,150 @@
+set_image_names() {
+  local filename_prefix
+  local -n os_image_name data_image_name
+  local -n boot_image_name verity_image_name root_image_name
```
I don't see why the caller needs to have these three names (boot/root/hash). `rpm2img` never interacts with files by these names because they're only created when compressed.

Also, `rpm2img` only needs `os_image_name` and `data_image_name` because the OVA creation change hasn't merged yet.

I would say, try to merge that one first, rebase, and remove the `set_image_names` and `get_data_image` functions until the caller is added, so I have more context to review.

jpculp replied:
Once everything is moved over, only `imghelper` will use them, but it felt kind of wrong to have `FILENAME_PREFIX="${IMAGE_NAME}-${VARIANT}-${ARCH}-${VERSION_ID}-${BUILD_ID}"` there. These values come from the docker environment calling `rpm2img` or `img2img`, but if we're ok with FILENAME_PREFIX (and the similar SYMLINK_PREFIX) being set at the top of `imghelper` it would make the code cleaner on both sides.

bcressey replied:
> if we're ok with FILENAME_PREFIX (and the similar SYMLINK_PREFIX) being set at the top of `imghelper` it would make the code cleaner on both sides.

OK, let's try it like that.

jpculp commented:
- Split extensions from ***_SUFFIX**.
- Adjusted wording on OVF error.
- Cleaned up `compress_image`.
- Unified `decompress_image`.
- Dropped `get_data_image` for a later PR.

jpculp committed:
rpm2img: move sanity checks to imghelper

Signed-off-by: Patrick J.P. Culp <jpculp@amazon.com>

bcressey commented on twoliter/embedded/imghelper:
```diff
@@ -1,6 +1,140 @@
+    tar xvf "${filename_prefix}.ova" "*${ext:+.${ext}}"
+    qemu-img convert -f "${ext}" -O raw "${filename_prefix}.vmdk" "${output_image}"
+    rm "${filename_prefix}${ext:+.${ext}}" "${filename_prefix}.ova"
+    ls -1
```
omit debugging commands:
```suggestion
```

bcressey commented on twoliter/embedded/imghelper:
```diff
@@ -1,6 +1,140 @@
+      "${input_dir}/${filename_prefix}${ext:+.${ext}}" "${output_image}"
+    ;;
+  vmdk)
+    tar xvf "${filename_prefix}.ova" "*${ext:+.${ext}}"
```
Why is the `*` needed?

jpculp replied:
This will extract multiple vmdks in the image. We can also just extract the whole tarball though.

jpculp replied:
This piece will be moved out of the decompress function, but for clarity it will be `"*.vmdk"` in `img2img`.

bcressey commented on twoliter/embedded/imghelper:
```diff
@@ -1,6 +1,140 @@
+    ;;
+  vmdk)
+    tar xvf "${filename_prefix}.ova" "*${ext:+.${ext}}"
+    qemu-img convert -f "${ext}" -O raw "${filename_prefix}.vmdk" "${output_image}"
+    rm "${filename_prefix}${ext:+.${ext}}" "${filename_prefix}.ova"
```
This won't handle split image layouts correctly - you are only extracting one file then removing the archive.

jpculp replied:
We're decompressing only the OS vmdk, but extracting all with `*${ext:+.${ext}}` on L98.

bcressey commented on twoliter/embedded/imghelper:
```diff
@@ -1,6 +1,140 @@
+  fi
+}
+
+decompress_image() {
+  local fmt input_dir filename_prefix output_image ext
```
I still think you need to align this with the `compress_image` signature.

Compress:
```
compress_image "vmdk" "os_image" "${OUTPUT_DIR}"
symlink_image "vmdk" "os_image" "${SYMLINK_PREFIX}" "${VERSION_ID}" "${OUTPUT_DIR}"
if [ -s "${DATA_IMAGE}" ] ; then
  compress_image "vmdk" "data_image" "${OUTPUT_DIR}"
  symlink_image "vmdk" "data_image" "${SYMLINK_PREFIX}" "${VERSION_ID}" "${OUTPUT_DIR}"
fi
```

Decompress:
```
decompress_image "vmdk" "os_image" "${OUTPUT_DIR}"
symlink_image "vmdk" "os_image" "${SYMLINK_PREFIX}" "${VERSION_ID}" "${OUTPUT_DIR}"
if [ -s "${DATA_IMAGE}" ] ; then
  decompress_image "vmdk" "data_image" "${OUTPUT_DIR}"
  symlink_image "vmdk" "data_image" "${SYMLINK_PREFIX}" "${VERSION_ID}" "${OUTPUT_DIR}"
fi
```

For that to work, you will want to leave the compressed artifact alone, so that the second call can extract the data image for the split VMDK case. But for consistency you should also not remove it in the raw or qcow2 cases.

jpculp committed:
rpm2img: move image compression to imghelper

Signed-off-by: Patrick J.P. Culp <jpculp@amazon.com>

jpculp committed:
imghelper: add decompression functions

Signed-off-by: Patrick J.P. Culp <jpculp@amazon.com>

jpculp commented:
- Move **FILENAME_PREFIX**, **SYMLINK_PREFIX**, and the **_IMAGE_NAME**s to `imghelper`
- Limit scope of `decompress_image` to make it more consistent with `compress_image`
- Remove `VERSION_ID` as an arg for `symlink_image` since we're already assuming it's in the environment and to make the inputs more consistent with `compress_image` and `decompress_image`.

jpculp commented:
Rebased.
